# iDEP and ShinyGO:   Graphical tools for analyzing genomics data  
## Let the data speak for itself!

This repository is now only host code for our webserver and some legacy scripts. GitHub repositories for [iDEP]x(https://github.com/gexijin/idepGolem) and [ShinyGO.](https://github.com/gexijin/shinygo).

[iDEP](http://bioinformatics.sdstate.edu/idep/) is a Shiny app for analyzing RNA-seq or other transcriptomic data. See [documentation](https://idepsite.wordpress.com/) and [paper](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-018-2486-6). Please visit the GitHub repository for the [new version of iDEP]x(https://github.com/gexijin/idepGolem)

iDEP's younger sister, [ShinyGO](http://bioinformatics.sdstate.edu/go/) focuses on the enrichment analysis of gene lists. The same database is used. ShinyGO is published [here](https://doi.org/10.1093/bioinformatics/btz931).  Please see separate repo for [ShinyGO.](https://github.com/gexijin/shinygo) Please report bugs or feature requests there.

## No Warranty
iDEP and ShinyGO are both under active development, and have not been fully tested. With the help of many users, we are constantly fixing bugs/errors in both source code and sometimes database. We recommend that you confirm your findings using other tools before publishing or further studies.


